### ISSUE TITLE
What is the best way to use a resolution that is not a multiple of 32?

### ISSUE THREAD

**Issue by deJQK** [ultralytics/ultralytics#20225]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am trying to train a model based on yolo code using my own models for backbone. However, it seems that the input resolution for yolo is always set to a multiple of 32, the stride used in the code. I tried to create my customized trainer and validator, together with YOLO model and detection model, to support a different stride to use an arbitrary input size, but it seems does not work as the output mAP is always zero. Currently I do the resizing inside my model, i.e., from say 640 to 610 in my model (the images are already square). I think as long as all the results only depends on relative size, say the loss and metrics does not depend on the original absolute image size, I think this method is reasonable. Is there any better solution for this? Thanks.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @deJQK, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to assist you faster 🕒. An Ultralytics engineer will review your question and provide further assistance soon!

**Comment by deJQK** [comment]
Thanks a lot @glenn-jocher for your kind reply and help. I tried the method you mentioned to adjust the parameters used in build_yolo_dataset. Indeed, I tried to customize the trainer, validator, YOLO model, as well as the detection model, to make sure using the customized stride (which is set to 1 in my case) wherever I noticed self.stride is defined or check_imgsz is called, to make sure not use the default stride of 32. However, although it can run, the results are not correct. I tried to debug on this, and seems the modified model of YOLO and detection models are reasonable (i.e., mAP, etc., are not zero). However, when using my private trainer (which also need to use a customized get_validator to use a validator with the modified stride value), the results of mAP are all zero. Do you have any idea on this? Also, why not to make stride a parameter that can be defined by the users... That is very easy for the framework and the default value of 32 can be always used (or maybe set it in the model config file) to make sure the behavior is proper for all the released models and checkpoints.

**Comment by glenn-jocher** [comment]
Setting stride=1 universally isn't recommended as it fundamentally alters how feature maps are processed in YOLO's architecture, which is designed around specific stride values (8/16/32) at different detection layers.

The zero mAP issue when using your private trainer likely stems from inconsistencies between how your custom components handle stride values. YOLO's detection mechanism depends on precise alignment between feature maps, anchors, and ground truth labels - all calculated based on consistent stride values.

A better approach would be:
1. Let the model use its native stride internally
2. Process your non-standard resolution images by either:
   - Padding to the nearest multiple of stride then removing padding at output
   - Using `rect=True` for validation which handles non-square images efficiently

This preserves the model's architecture while accommodating your specific resolution requirements. For training with irregular image sizes, rectangular training is often more efficient than forcing all images to an arbitrary dimension.

**Comment by deJQK** [comment]
Thanks @glenn-jocher .

For the method you mentioned "padding to the neareast multiple of stride", how could I do that? In the original code of yolo, it uses the dataset YOLODataset, which use v8transforms and LetterBox for the image. If I would like to use a pretransform of padding on my images (and also labels), it seems I need to implement a child class of YOLODataset to use my custom pretransforms. It seems that I should not do the padding inside my model itself.

On the other hand, if I can guarantee the input batch after data processing by the yolo code is a square, say it is converted to 640, could I directly interpolate/resize on my input batch to the desired shape 610 and run my model? All the outputs of the model and labels are relative size, which does not change if I only scale/resize it when keeping the image square, right? Or if I miss some point for the output postprocessing using this method?

Thanks.

**Comment by glenn-jocher** [comment]
You have two options for handling non-standard resolutions like 610:

1. For the padding approach, you don't need to create a custom YOLODataset. Instead, you can use the existing LetterBox transform with custom parameters:

```python
from ultralytics.data.augment import LetterBox

# In your preprocessing pipeline
transform = LetterBox(new_shape=(610, 610), auto=False, stride=1)
# This will pad to your desired size rather than to a multiple of stride
```

2. For the resizing approach after YOLO preprocessing: Yes, you can resize from 640→610 if:
   - You maintain the square aspect ratio
   - Your model handles the new size consistently
   - You apply the same resize factor to annotations during validation

However, this approach may introduce subtle alignment issues between predictions and ground truth. The more standard approach is to let YOLO handle the letterboxing with rect=True and focus on building your custom backbone without changing the stride mechanism.

**Comment by deJQK** [comment]
Thanks a lot again @glenn-jocher .

For the first part, after I implement my customized LetterBox, how could I integrate that into the YOLO model? For example, I have
```
model = YOLO(model_cfg)
```
where model_cfg defines the custom backbones and heads. Where could I insert the implemented LetterBox into the model to use it, instead of the original one?

For the second, I can guarantee the first two. For "apply the same resize factor to annotations during validation", do you mean on the images and labels? If the model's final output (from the Detect head) is relative position of bboxes, why do I need to resize labels?

Regarding `rect=True`, during evaluation, the model uses LetterBox, which will convert the input again to a square by padding and resizing, even without specifying `rect`. If so, why should I take care of that during evaluation?

Thanks!

**Comment by glenn-jocher** [comment]
@deJQK The LetterBox transform is applied within the dataset class, not directly in the YOLO model. To use your custom LetterBox implementation, you'd need to override the `build_transforms` method in a custom dataset class that inherits from YOLODataset.

Regarding resizing labels: While YOLO uses relative coordinates, the validation process compares predictions against ground truth annotations. When you resize images, the model's internal feature map sizing changes, which can affect how predictions align with ground truth, potentially causing mAP discrepancies.

As for rect=True, it specifically tells the validator to batch similar aspect ratio images together and resize efficiently, which is different from the standard letterboxing. This helps maintain proper aspect ratios during validation while optimizing memory usage.

If you're primarily trying to use a non-standard resolution like 610, consider using check_imgsz with a custom floor parameter to maintain your desired dimensions:
```python
adjusted_size = check_imgsz(610, stride=32, floor=610)
```

**Comment by deJQK** [comment]
Thanks a lot @glenn-jocher again for your kind and patient explanation. Unfortunately, it might not be that easy as you just mentioned.

If I would like to use the `check_imgsz` as you mentioned, many different places must be modified (which in turn means inherited classes need to be implemented), suppose we would like to run the training and evaluation as usual, i.e.,
```
model = YOLO(model_cfg)
model.train(task='detect', mode='train', device=0, imgsz=610)
metrics = model.val(save_json=True)
print(metrics)
```

Here are the points that call `check_imgsz` internally and changing the `imgsz` to multiple of the `stride`, or changing the `stride` internally:

Firstly, the model definition command `model = YOLO(model_cfg)` will call the `task_map` function ([here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/model.py#L66)), which in turn will call the predefined model `DetectionModel` from `ultralytics.nn.tasks` defined [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/tasks.py#L308). In this model's method `__init__` [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/tasks.py#L342-L356), it will run my model with some resolution 256, which is undesired and different from 610. So to overcome this, a children class must be implemented to modify the behavior here.

Next, the model definition described above calling the `task_map` function will call the predefined trainer `DetectionTrainer` from `ultralytics.models.yolo.detect.train` defined [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/detect/train.py#L19). In this trainer's method `build_dataset` [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/detect/train.py#L52), `stride` is modified to 32 [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/detect/train.py#L64), and in the method `get_model` defined [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/detect/train.py#L128), the customized `DetectionModel` with the proposed `stride=1` need to be called. The `get_validator` method [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/detect/train.py#L145) also needs to be customized to call a new validator with the new `stride` value. Also the method `_setup_train` from the parent class `BaseTrainer` need to be customized, where at [these lines](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/trainer.py#L280-L282), the stride is internally modified to 32, not the custom values.

For the validator `DetectionValidator` from `ultralytics.models.yolo.detect.val`, which is called also by the `task_map` function mentioned above, the `__init__` and `__call__` methods from the base class `BaseValidator` defined [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/validator.py#L42). Specifically, [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/validator.py#L126), `check_imgsz` is also called [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/validator.py#L126) with the default value of `stride=32`. `stride` is also modified to 32 in [this line](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/validator.py#L170) by the class `AutoBackend`, which also need to be adjusted.

Even with all the above definitions of children classes of `DetectionModel`, `DetectionTrainer`, `DetectionValidator`, as well as `YOLO`, the final mAP is still vanishing, together with other metrics. I am not sure where I missed to modify.

Thanks again.

**Comment by glenn-jocher** [comment]
@deJQK You're hitting a fundamental design aspect of YOLO - the stride value of 32 is deeply integrated throughout the architecture. This isn't just a simple parameter you can change; it's tied to how feature maps are processed at different scales.

Based on the reference materials, particularly the `LetterBox` class implementation, I'd recommend:

1. Instead of trying to override the stride system completely, use the existing letterboxing mechanism but with custom parameters:
   ```python
   # In your validator's preprocess method
   from ultralytics.data.augment import ClassifyLetterBox
   transform = ClassifyLetterBox(size=610, auto=False)
   ```

2. For training with non-standard sizes, enable rectangular training with `rect=True` which handles variable image dimensions more gracefully.

The reason your mAP is vanishing despite all the modifications is likely due to misalignment between predictions and ground truth caused by inconsistent stride handling across the detection pipeline. The letterboxing and NMS processes are carefully tuned to work with specific strides.

If you absolutely must use 610px, consider accepting the small padding overhead and continuing to use the native stride mechanisms internally.

**Comment by deJQK** [comment]
Thanks @glenn-jocher . If I pad my input from 610 to a nearest multiple of 32, i.e., 640 (i.e. all borders are padded with background), and then center crop the input from 640 to 610 inside my model, and use this to run with the yolo codebase, would the resulting metrics correct? Are the iou, map, etc. calculated based in input resolution of 640, or input resolution of 610? Thanks!

**Comment by deJQK** [comment]
Also, I noticed that, the imgsz for train and valid can be different, although my model is defined and imgsz is prescribed for train on the imgsz. Why can this happen?

**Comment by deJQK** [comment]
I noticed that this might be general. For example, if my code is something like this:

```
model = YOLO(model_yaml_file_path)
model.cuda()
model.train(
  task='detect',
  mode='train',
  project=output_dir,
  name='train',
  device=0,
  data=data_yaml_file_path,
  model=model_yaml_filename,
  epochs=1,
  batch=16,
  patience=0,
  imgsz=640,
  mosaic=0.0,
  freeze=0)
model.val(
  project=output_dir,
  name='valid',
  save_json=True,
  imgsz=640)
```

In the first call of model.train, the size of the input to the model is [bs, 3, 640, 640] during training, but for the eval or final_eval for each epoch, the size becomes [bs, 3, 672, 672]. This also happen for the model.val and the size is [bs, 3, 672, 672]. Why does this happen?

**Comment by glenn-jocher** [comment]
@deJQK The different image sizes you're seeing between training and validation are due to the `rect=True` parameter that's enabled by default in validation mode. Looking at the reference code, this creates rectangular batches based on aspect ratios for more efficient processing.

From the validator source code, when `rect=True` (default for validation), it applies rectangular inference batching which can result in image sizes that differ from your specified `imgsz`, often rounding up to the next multiple of stride.

The best way to handle a non-standard resolution like 610px is:

1. Accept the automatic padding during validation, as metrics are calculated based on relative coordinates and won't be affected by the padding

2. If you must use exactly 610px, modify the `LetterBox` transform directly:
   ```python
   # Create a custom validation dataset with fixed size
   from ultralytics.data.augment import LetterBox
   custom_letterbox = LetterBox(new_shape=(610, 610), auto=False, scale_fill=True)
   ```

3. Set `rect=False` in your validation parameters to maintain consistent image sizing:
   ```python
   model.val(imgsz=610, rect=False)
   ```

The metrics are calculated using relative coordinates on the processed images, so as long as the aspect ratio is maintained, your evaluation should be accurate regardless of the actual pixel dimensions.

**Comment by deJQK** [comment]
Thanks @glenn-jocher . I tried to use
```
model.val(
  project=output_dir,
  name='valid',
  save_json=True,
  rect=False,
  imgsz=640)
```
but still get 672 as input size for the model.

**Comment by glenn-jocher** [comment]
@deJQK Despite setting `rect=False`, you're still seeing 672 because there are multiple places in the pipeline where image size adjustments occur. The most likely culprit is the `AutoBackend` model loading which applies its own size calculations.

To force exactly the size you want, you can try a more direct approach:

```python
# Create validator with explicit parameters
from ultralytics.engine.validator import DetectionValidator
validator = DetectionValidator(args=dict(
    imgsz=640,
    rect=False,
    stride=32,
    pad=0,  # Try with minimal padding
    data=data_yaml_path
))
results = validator(model=model)
```

If you're working with a custom backbone that truly requires a specific non-multiple-of-32 resolution, consider implementing a custom preprocessing hook that respects your exact dimensions rather than modifying the entire stride system. This would be more maintainable than trying to override all the places where YOLO applies size adjustments.
